### PR TITLE
AWS SDK 3x compatibility

### LINKED ISSUE #827: Configuration failure
Seems new AWS SDK is not supported:
aws/aws-sdk-php (3.2.6)

```
Missing required client configuration options: 

region: (string)

A "region" configuration value is required for the "s3" service
(e.g., "us-west-2"). A list of available public regions and endpoints can be
found at http://docs.aws.amazon.com/general/latest/gr/rande.html.

version: (string)

A "version" configuration value is required. Specifying a version constraint
ensures that your code will not be affected by a breaking change made to the
service. For example, when using Amazon S3, you can lock your API version to
"2006-03-01".

Your build of the SDK has the following version(s) of "s3": * "2006-03-01"

You may provide "latest" to the "version" configuration value to utilize the
most recent available API version that your client's API provider can find.
Note: Using 'latest' in a production application is not recommended.

A list of available API versions can be found on each client's API documentation
page: http://docs.aws.amazon.com/aws-sdk-php/v3/api/index.html. If you are
unable to load a specific API version, then you may need to update your copy of
the SDK.
```

### PR DESCRIPTION
**Pull request by nicolasricci**
Closes #827 

### Changelog

```markdown
### Added
- Add parameter to specify aws_sdk version

### Changed
- Update configuration and metadatabuilder to comply with AWS SDK 3.x
```

### Subject

This PR allow you to use the AWS SDK 3.x branch for S3 management

### CODE REVIEW

**Comment by OskarStark** [comment]
minor or patch? @sonata-project/contributors

File: Metadata/AmazonMetadataBuilder.php [hunk 1]
```diff
@@ -11,13 +11,22 @@
...
+    const PRIVATE_ACCESS = 'private';
+    const PUBLIC_READ = 'public-read';
+    const PUBLIC_READ_WRITE = 'public-read-write';
+    const AUTHENTICATED_READ = 'authenticated-read';
+    const BUCKET_OWNER_READ = 'bucket-owner-read';
+    const BUCKET_OWNER_FULL_CONTROL = 'bucket-owner-full-control';
+
+    const STORAGE_STANDARD = 'STANDARD';
+    const STORAGE_REDUCED = 'REDUCED_REDUNDANCY';
+    const STORAGE_GLACIER = 'GLACIER';
```
Comment (greg0ire) [inline Metadata/AmazonMetadataBuilder.php]: @OskarStark new things : at least minor

**Comment by nicolasricci** [comment]
@OskarStark is it normal that travis is constantly failing? not sure how to fix this?

File: DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -215,6 +215,8 @@ private function addFilesystemSection(ArrayNodeDefinition $node)
                                     ->end()
                                 ->end()
                                 ->scalarNode('region')->defaultValue('s3.amazonaws.com')->end()
+                                ->scalarNode('version')->defaultValue('latest')->end()
+                                ->scalarNode('sdk_version')->defaultValue(2)->end()
```
Comment (Soullivaneuh) [inline DependencyInjection/Configuration.php]: Can you limit the possible values?

http://symfony.com/doc/current/components/config/definition.html#enum-nodes

File: DependencyInjection/SonataMediaExtension.php [hunk 3]
```diff
@@ -379,13 +379,25 @@ public function configureFilesystemAdapter(ContainerBuilder $container, array $c
                 ))
             ;
 
-            $container->getDefinition('sonata.media.adapter.service.s3')
+            if ($config['filesystem']['s3']['sdk_version'] === 3) {
```
Comment (Soullivaneuh) [inline DependencyInjection/SonataMediaExtension.php]: Yoda condition needed.

http://symfony.com/doc/current/contributing/code/standards.html#structure

File: Metadata/AmazonMetadataBuilder.php [hunk 4]
```diff
@@ -27,12 +36,12 @@ class AmazonMetadataBuilder implements MetadataBuilderInterface
      * @var string[]
      */
     protected $acl = array(
-        'private' => CannedAcl::PRIVATE_ACCESS,
-        'public' => CannedAcl::PUBLIC_READ,
-        'open' => CannedAcl::PUBLIC_READ_WRITE,
-        'auth_read' => CannedAcl::AUTHENTICATED_READ,
-        'owner_read' => CannedAcl::BUCKET_OWNER_READ,
-        'owner_full_control' => CannedAcl::BUCKET_OWNER_FULL_CONTROL,
+        'private' => self::PRIVATE_ACCESS,
```
Comment (Soullivaneuh) [inline Metadata/AmazonMetadataBuilder.php]: Could you use `static` instead of `self`?

File: Metadata/AmazonMetadataBuilder.php [hunk 5]
```diff
@@ -70,9 +79,9 @@ protected function getDefaultMetadata()
         //merge storage
         if (isset($this->settings['storage'])) {
             if ($this->settings['storage'] == 'standard') {
-                $output['storage'] = Storage::STANDARD;
+                $output['storage'] = self::STORAGE_STANDARD;
```
Comment (Soullivaneuh) [inline Metadata/AmazonMetadataBuilder.php]: static

File: Metadata/AmazonMetadataBuilder.php [hunk 6]
```diff
@@ -70,9 +79,9 @@ protected function getDefaultMetadata()
...
             } elseif ($this->settings['storage'] == 'reduced') {
-                $output['storage'] = Storage::REDUCED;
+                $output['storage'] = self::STORAGE_REDUCED;
```
Comment (Soullivaneuh) [inline Metadata/AmazonMetadataBuilder.php]: static

File: Tests/Metadata/AmazonMetadataBuilderTest.php [hunk 7]
```diff
@@ -42,18 +41,18 @@ public function testAmazon()
     public function provider()
     {
         return array(
-            array(array('acl' => 'private'), array('ACL' => CannedAcl::PRIVATE_ACCESS, 'contentType' => 'image/png')),
```
Comment (Soullivaneuh) [inline Tests/Metadata/AmazonMetadataBuilderTest.php]: Why those tests are removed?

**Comment by Soullivaneuh** [comment]
> @OskarStark is it normal that travis is constantly failing? not sure how to fix this?

Indeed 3.x is failing. Will take a look.

**Comment by Soullivaneuh** [comment]
@nicolasricci see: https://github.com/sonata-project/SonataCoreBundle/pull/293

I'll release a new version of core-bundle ASAP after that.

**Comment by Soullivaneuh** [comment]
@nicolasricci Fixed, test will pass. :+1:

**Comment by nicolasricci** [comment]
thanks @Soullivaneuh for the feedback. I will make the corrections in the evening

File: Tests/Metadata/AmazonMetadataBuilderTest.php (hunk 7 above)
Comment (nicolasricci) [inline Tests/Metadata/AmazonMetadataBuilderTest.php]: they have not been removed just updated to use the correct constant

File: Metadata/AmazonMetadataBuilder.php (hunk 4 above)
Comment (nicolasricci) [inline Metadata/AmazonMetadataBuilder.php]: using static would result in `static:: is not allowed in compile-time constants`

**Comment by nicolasricci** [comment]
@Soullivaneuh can this be please be reviewed  for merge now that all tests are green

File: Resources/doc/reference/amazon_s3.rst [hunk 8]
```diff
@@ -22,7 +22,13 @@ This is a sample config file to enable amazon S3 as a filesystem & provider:
 
         filesystem:
             s3:
-                bucket:    %s3_bucket_name%
-                accessKey: %s3_access_key%
-                secretKey: %s3_secret_key%
-                region:    %s3_region%
+                bucket:      %s3_bucket_name%
```
Comment (OskarStark) [inline Resources/doc/reference/amazon_s3.rst]: should be surrounded by `"` i would say

File: Tests/Metadata/AmazonMetadataBuilderTest.php (hunk 7 above)
Comment (OskarStark) [inline Tests/Metadata/AmazonMetadataBuilderTest.php]: thats true 👍

**Comment by OskarStark** [comment]
LGTM 👍

**Comment by nicolasricci** [comment]
ping @Soullivaneuh

**Comment by nicolasricci** [comment]
@Soullivaneuh we are soon launching a project using this PR and it would nice if it could be merged so I don't have to tweak the composer with my fork and keep the SonataMediaBundle up-to-date.

**Comment by OskarStark** [comment]
We should squash the commits imo

/cc @greg0ire @core23 @Soullivaneuh

**Comment by greg0ire** [comment]
Maybe not all, but some of them for sure. @nicolasricci , can you use `git rebase origin/3.x` to get rid of the merge commits, then `git rebase --interactive origin/3.x` to do some cleanup ?

**Comment by nicolasricci** [comment]
@greg0ire sorry for the late reply. i have rebased my branch but the squash did not work properly it seems. is it an issue?

**Comment by greg0ire** [comment]
> @greg0ire sorry for the late reply. i have rebased my branch but the squash did not work properly it seems. is it an issue?

O_o indeed. Try rebasing again, this time with the `--interactive` option. You should see a list of commits. Just remove those that are not yours.

**Comment by nicolasricci** [comment]
That's what I did but I thought you wanted me to squash the commits. All those listed in the `--interactive` mode are mine . S

```
pick 8d50e68 add parameter to specify aws_sdk version update configuration and metadatabuilder to comply with aws sdk 3.x
pick f656ebd fix use of constants php cs
pick a118114 remove unused arguments
pick cf3668b add doc fix cs
pick 14feb0c remove minimum-stability and prefer-stable parameters
pick 4145db4 compatibility changes for aws-php-sdk 3.x
pick e4896dd add parameter to set aws sdk version
pick ce92692 rename parameter and fix argument injection
pick 58441fe remove duplicated config fix cs
pick 3ea6395 remove minimum-stability and prefer-stable parameters
pick 1ff3a69 remove unused use statement
pick cfa0d17 fix cs
pick 2c2494d fix cs
pick 28c8c34 change sdk_version configuration to enum_node use static instead of const when possible yoda condition
```

but when I try to apply the rebase I have the following error

```
error: could not apply 8d50e68... add parameter to specify aws_sdk version

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply 8d50e68f86e463a8e43e33ffeb8dcdee0a62c210... add parameter to specify aws_sdk version update configuration and metadatabuilder to comply with aws sdk 3.x
```

Any idea how to solve it?

**Comment by greg0ire** [comment]
> Any idea how to solve it?

This is a conflict. Type `git mergetool` and fix it, then continue your rebase with `git rebase --continue`

**Comment by nicolasricci** [comment]
OK my fork was so out of sync that I had to `rebase --continue` multiple times before it succeeded but it seems OK now
`Successfully rebased and updated refs/heads/aws-sdk-3x-compatibility.`

**Comment by greg0ire** [comment]
It looks way better indeed. You should try to learn rebase interactive to fixup meaningless commits into meaningful commits (I'm speaking about the commits starting with "fix")

**Comment by nicolasricci** [comment]
something wrong in the merge, I have pushed a fix. the `rebase --interactive` does not bring anything up now that i have pushed. how can i squash all those "fix" commits?

**Comment by greg0ire** [comment]
If you want to see all the commits since commit A, use `git rebase --interactive A`. The commits can be squashed or fixed up by replacing the `pick` instruction in the list by `squash` (or `s`) or `fixup` (or `f`), respectively. The commit will be squashed or fixup into the previous commit, so make sure you move the line corresponding to the commit under the line corresponding to the commit you want to squash into.

**Comment by nicolasricci** [comment]
👍 thanks for the git tips ;) i have now squashed all the irrelevant commits

**Comment by greg0ire** [comment]
Looks a lot cleaner :)

**Comment by nicolasricci** [comment]
good to finally merge then or do you need me to do anything else?

File: DependencyInjection/SonataMediaExtension.php [hunk 9]
```diff
@@ -428,7 +427,6 @@ public function configureFilesystemAdapter(ContainerBuilder $container, array $c
                     'version' => $config['filesystem']['s3']['version'],
                 ))
             ;
-
```
Comment (greg0ire) [inline DependencyInjection/SonataMediaExtension.php]: This commit should be fixed up with the previous one : we don't want to see in the history your first try at formatting this.

File: DependencyInjection/Configuration.php [hunk 10]
```diff
@@ -216,7 +216,10 @@ private function addFilesystemSection(ArrayNodeDefinition $node)
                                 ->end()
                                 ->scalarNode('region')->defaultValue('s3.amazonaws.com')->end()
                                 ->scalarNode('version')->defaultValue('latest')->end()
+<<<<<<< HEAD
                                 ->scalarNode('sdk_version')->defaultValue(2)->end()
+=======
+>>>>>>> 4145db4... compatibility changes for aws-php-sdk 3.x
```
Comment (greg0ire) [inline DependencyInjection/Configuration.php]: Wow. This should never ever appear in a commit.

File: DependencyInjection/Configuration.php [hunk 11]
```diff
@@ -216,10 +216,7 @@ private function addFilesystemSection(ArrayNodeDefinition $node)
                                 ->end()
                                 ->scalarNode('region')->defaultValue('s3.amazonaws.com')->end()
                                 ->scalarNode('version')->defaultValue('latest')->end()
-<<<<<<< HEAD
-                                ->scalarNode('sdk_version')->defaultValue(2)->end()
-=======
->>>>>>> 4145db4... compatibility changes for aws-php-sdk 3.x
```
Comment (greg0ire) [inline DependencyInjection/Configuration.php]: neither should this.

File: DependencyInjection/Configuration.php [hunk 12]
```diff
@@ -216,7 +216,7 @@ private function addFilesystemSection(ArrayNodeDefinition $node)
                                 ->end()
                                 ->scalarNode('region')->defaultValue('s3.amazonaws.com')->end()
                                 ->scalarNode('version')->defaultValue('latest')->end()
-                                ->scalarNode('aws_sdk_version')->defaultValue(2)->end()
+                                ->scalarNode('sdk_version')->defaultValue(2)->end()
```
Comment (greg0ire) [inline DependencyInjection/Configuration.php]: Same here. We don't need to see that, fix it up into the commit that introduced this.

**Comment by greg0ire** [comment]
You're going to have a hard time unless you simply redo the needed commit and only them. So : 

1. Forget all your commits, and keep the changes : `git reset --mixed 3.x`
2. Use `git add --patch` to prepare a commit with the changes you want
3. Use `git commit` to give that set of changes a meaningful commit message
4. Use `git push --force` to replace the commits that we can see here with your new, beautiful commits.

**Comment by antoine64** [comment]
Can this PR be accepted ? Need this for AWS working with latests aws-sdk-php versions.

**Comment by greg0ire** [comment]
I'm retrying the failing job.

**Comment by greg0ire** [comment]
Yaaay, it's passing!

File: Tests/Metadata/AmazonMetadataBuilderTest.php [hunk 13]
```diff
@@ -42,18 +41,18 @@ public function testAmazon()
...
             array(array('cache_control' => 'max-age=86400'), array('CacheControl' => 'max-age=86400', 'contentType' => 'image/png')),
             array(array('encryption' => 'aes256'), array('encryption' => 'AES256', 'contentType' => 'image/png')),
             array(array('meta' => array('key' => 'value')), array('meta' => array('key' => 'value'), 'contentType' => 'image/png')),
-            array(array('acl' => 'public', 'storage' => 'standard', 'cache_control' => 'max-age=86400', 'encryption' => 'aes256', 'meta' => array('key' => 'value')), array('ACL' => CannedAcl::PUBLIC_READ, 'storage' => Storage::STANDARD, 'meta' => array('key' => 'value'), 'CacheControl' => 'max-age=86400', 'encryption' => 'AES256', 'contentType' => 'image/png')),
+            array(array('acl' => 'public', 'storage' => 'standard', 'cache_control' => 'max-age=86400', 'encryption' => 'aes256', 'meta' => array('key' => 'value')), array('ACL' => AmazonMetadataBuilder::PUBLIC_READ, 'storage' => Storage::STANDARD, 'meta' => array('key' => 'value'), 'CacheControl' => 'max-age=86400', 'encryption' => 'AES256', 'contentType' => 'image/png')),
```
Comment (greg0ire) [inline Tests/Metadata/AmazonMetadataBuilderTest.php]: Uuuh… do you guys know about a thing called "line breaks" ?

Comment (greg0ire) [inline Tests/Metadata/AmazonMetadataBuilderTest.php]: Merging it since @antoine64 is waiting, but this should get fixed one day, if possible :P

**Pull request merged**